File: annotation-processor/src/main/java/me/bristermitten/mittenlib/annotations/compile/SerializationCodeGenerator.java
package me.bristermitten.mittenlib.annotations.compile;

import com.google.inject.Inject;
import com.palantir.javapoet.ClassName;
import com.palantir.javapoet.CodeBlock;
import com.palantir.javapoet.MethodSpec;
import com.palantir.javapoet.ParameterSpec;
import com.palantir.javapoet.ParameterizedTypeName;
import com.palantir.javapoet.TypeName;
import com.palantir.javapoet.TypeSpec;
import com.palantir.javapoet.WildcardTypeName;
import io.toolisticon.aptk.tools.TypeMirrorWrapper;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import javax.lang.model.element.Modifier;
import javax.lang.model.element.TypeElement;
import javax.lang.model.type.DeclaredType;
import javax.lang.model.type.TypeMirror;
import me.bristermitten.mittenlib.annotations.ast.AbstractConfigStructure;
import me.bristermitten.mittenlib.annotations.ast.CustomSerializerInfo;
import me.bristermitten.mittenlib.annotations.ast.Property;
import me.bristermitten.mittenlib.annotations.parser.CustomSerializers;
import me.bristermitten.mittenlib.annotations.util.NewtypeUtil;
import me.bristermitten.mittenlib.annotations.util.TypesUtil;
import me.bristermitten.mittenlib.config.CollectionsUtils;
import me.bristermitten.mittenlib.config.SerializationContext;
import me.bristermitten.mittenlib.config.extension.UseObjectMapperSerialization;
import me.bristermitten.mittenlib.config.tree.DataTree;
import me.bristermitten.mittenlib.config.tree.DataTreeTransforms;
import me.bristermitten.mittenlib.util.Strings;

/**
 * Generates serialization code for configuration classes. This class is responsible for creating
 * methods that convert from strongly typed configuration objects to {@link DataTree}
 * representations.
 */
public class SerializationCodeGenerator {

    private final ConfigurationClassNameGenerator configurationClassNameGenerator;
    private final TypesUtil typesUtil;
    private final CustomSerializers customSerializers;
    private final MethodNames methodNames;

    @Inject
    public SerializationCodeGenerator(
            ConfigurationClassNameGenerator configurationClassNameGenerator,
            TypesUtil typesUtil,
            CustomSerializers customSerializers,
            MethodNames methodNames) {
        this.configurationClassNameGenerator = configurationClassNameGenerator;
        this.typesUtil = typesUtil;
        this.customSerializers = customSerializers;
        this.methodNames = methodNames;
    }

    /**
     * Checks if serialization is fully supported for the given configuration. Serialization is
     * supported if all properties either: - Are natively supported types - Are @Config types -
     * Have @UseObjectMapperSerialization annotation
     *
     * <p>Properties with CustomDeserializers that don't have serialization support will prevent full
     * serialization from being generated.
     *
     * @param ast The configuration structure to check
     * @return true if serialization can be fully generated, false otherwise
     */
    public boolean isSerializationSupported(AbstractConfigStructure ast) {
        for (Property property : ast.properties()) {
            if (propertyIsUnserializable(property)) {
                return false;
            }
        }
        return true;
    }

    /**
     * Gets a list of properties that cannot be serialized. Useful for generating helpful
     * error/warning messages.
     *
     * @param ast The configuration structure to check
     * @return List of property names that cannot be serialized
     */
    public List<String> getUnsupportedSerializationProperties(AbstractConfigStructure ast) {
        List<String> unsupported = new ArrayList<>();
        for (Property property : ast.properties()) {
            if (propertyIsUnserializable(property)) {
                unsupported.add(property.name() + " (" + property.propertyType() + ")");
            }
        }
        return unsupported;
    }

    /** Checks if a single property can be serialized. */
    private boolean propertyIsUnserializable(Property property) {
        // explicitly marked as using ObjectMapper serialization - always serializable
        if (typesUtil.getAnnotation(property.source().element(), UseObjectMapperSerialization.class) != null) {
            return false;
        }

        TypeMirror propertyTypeMirror = property.propertyType();
        TypeMirrorWrapper wrappedType = TypeMirrorWrapper.wrap(propertyTypeMirror);

        // Check generic types
        if (wrappedType.hasTypeArguments()) {
            String canonicalName = wrappedType.erasure().getQualifiedName();

            if (typesUtil.isCollection(propertyTypeMirror) || canonicalName.equals(Map.class.getName())) {
                var typeArguments = wrappedType.getTypeArguments();
                for (TypeMirror typeArgument : typeArguments) {
                    if (propertyIsUnserializable(
                            new Property(property.name(), typeArgument, property.source(), property.settings()))) {
                        return true;
                    }
                }
                return false;
            }
            // Unknown generic type: treat as unserializable unless explicitly opted in via UseObjectMapperSerialization
            return true;
        }

        // Config types are always serializable
        if (typesUtil.isConfigType(propertyTypeMirror)) {
            return false;
        }

        // Newtypes are serializable if their underlying type is serializable
        if (typesUtil.isNewtype(propertyTypeMirror)) {
            TypeElement typeElement = (TypeElement) ((DeclaredType) propertyTypeMirror).asElement();
            TypeMirror underlyingType = NewtypeUtil.getUnderlyingType(typeElement);
            return propertyIsUnserializable(
                    new Property(property.name(), underlyingType, property.source(), property.settings()));
        }

        if (isKnownSerializableType(wrappedType)) {
            return false;
        }

        // Unknown type - only serializable if CustomSerializer is present
        return customSerializers.getCustomInfo(propertyTypeMirror).isEmpty();
    }

    /**
     * Adds private serialization methods for each property of a configuration structure to its saver
     * class.
     *
     * @param typeSpecBuilder the builder for the saver class
     * @param ast the configuration structure
     */
    public void addSerializeMethodsToSaver(TypeSpec.Builder typeSpecBuilder, AbstractConfigStructure ast) {
        // Generate serialize methods for each property
        for (Property property : ast.properties()) {
            MethodSpec serializeMethod = createSerializeMethodFor(property);
            typeSpecBuilder.addMethod(serializeMethod);
        }
    }

    /**
     * Get the parameter type for a serialization method. This method adds wildcards to collection
     * types to allow for both the public and implementation types.
     *
     * <p>For example, a property with type {@code List<T>} will be mapped to {@code List<? extends
     * T>}
     */
    private TypeName getSerializeParameterType(Property property) {
        TypeName typeName = configurationClassNameGenerator.publicPropertyClassName(property);
        if (typeName instanceof ParameterizedTypeName parameterizedTypeName) {
            ClassName rawType = parameterizedTypeName.rawType();
            if (rawType.equals(ClassName.get(List.class))
                    || rawType.equals(ClassName.get(Set.class))
                    || rawType.equals(ClassName.get(Map.class))) {
                List<TypeName> typeArguments = parameterizedTypeName.typeArguments().stream()
                        .map(arg -> (TypeName) WildcardTypeName.subtypeOf(arg))
                        .toList();
                return ParameterizedTypeName.get(rawType, typeArguments.toArray(new TypeName[0]));
            }
        }
        return typeName;
    }

    /**
     * Creates a private serialization method for a specific property.
     *
     * <p>Generates:
     *
     * <pre>{@code
     * private DataTree serializeCount(Integer value, SerializationContext context) {
     *     if (value == null) return DataTree.null_();
     *     DataTree result;
     *     result = DataTreeTransforms.loadFrom(value);
     *     return result;
     * }
     * }</pre>
     *
     * @param property the property to create a serialization method for, used for the method name
     *     (e.g. {@code serializeCount}) and return type
     * @return a method spec for the serialization method
     */
    private MethodSpec createSerializeMethodFor(Property property) {
        String methodName = methodNames.getSerializeMethodName(property);
        TypeName propertyType = getSerializeParameterType(property);

        // Check if the property is annotated with @UseObjectMapperSerialization
        boolean useObjectMapper =
                typesUtil.getAnnotation(property.source().element(), UseObjectMapperSerialization.class) != null;

        MethodSpec.Builder builder = MethodSpec.methodBuilder(methodName)
                .addJavadoc("""
                        Serializes the {@code $L} property into a {@link $T}.

                        @param value the value to serialize
                        @param context the serialization context
                        @return the serialized DataTree representation
                        """, property.name(), DataTree.class)
                .addModifiers(Modifier.PRIVATE)
                .returns(DataTree.class)
                .addParameter(ParameterSpec.builder(propertyType, "value").build())
                .addParameter(ParameterSpec.builder(SerializationContext.class, "context")
                        .build());

        // Handle null values
        if (property.settings().isNullable()) {
            builder.beginControlFlow("if (value == null)");
            builder.addStatement("return $T.null_()", DataTree.class);
            builder.endControlFlow();
        }

        if (useObjectMapper) {
            // Use ObjectMapper for serialization
            handleObjectMapperSerialization(builder);
            return builder.build();
        }

        builder.addStatement("$T result", DataTree.class);
        builder.addCode(generateSerialization(property.propertyType(), "value", "result", 0));
        builder.addStatement("return result");

        return builder.build();
    }

    /** Checks if a type is a known serializable type (primitives, String, Boolean, Number, enums). */
    private boolean isKnownSerializableType(TypeMirrorWrapper wrappedType) {
        if (wrappedType.getTypeElement().isEmpty() && wrappedType.isPrimitive()) {
            return true;
        }

        if (typesUtil.getDataTreeType(TypeName.get(wrappedType.unwrap())).isPresent()) {
            return true;
        }

        if (wrappedType
                .getQualifiedName()
                .equals(Character.class.getName())) { // TODO: do we actually know how to support Character?
            return true;
        }
        return wrappedType.isEnum();
    }

    /**
     * Handles serialization by delegating to an {@link
     * me.bristermitten.mittenlib.config.reader.ObjectMapper}.
     *
     * <p>Generates:
     *
     * <pre>{@code
     * return DataTreeTransforms.loadFrom(context.getMapper().map(value));
     * }</pre>
     *
     * @param builder the method builder
     */
    private void handleObjectMapperSerialization(MethodSpec.Builder builder) {
        // Use ObjectMapper to map the value and then load it as DataTree
        // We use mapper.map(value) which returns an Object (likely a Map or List)
        // then pass that to DataTreeTransforms.loadFrom
        builder.addStatement("return $T.loadFrom(context.getMapper().map(value))", DataTreeTransforms.class);
    }

    /**
     * Recursively generates serialization {@link CodeBlock} statements for a given type, converting
     * it to a {@link DataTree} representation. Supports custom serializers, nested config types,
     * collections ({@link List}, {@link Map}), and primitive/built-in serializable types.
     *
     * <p>For a {@link List}, generates:
     *
     * <pre>{@code
     * DataTree[] arr0 = new DataTree[value.size()];
     * for (int i0 = 0; i0 < value.size(); i0++) {
     *     String el0 = (String) value.get(i0);
     *     arr0[i0] = DataTreeTransforms.loadFrom(el0);
     * }
     * result = DataTree.array(arr0);
     * }</pre>
     *
     * @param type the type of the element being serialized
     * @param inputVar the name of the local variable holding the value to serialize (e.g. {@code
     *     value})
     * @param targetExpression the code expression to assign the resulting {@link DataTree} to (e.g.
     *     {@code result} or {@code arr0[i0]})
     * @param depth the current recursion depth, used to generate unique variable names and avoid
     *     local variable scope clashes (e.g. {@code arr0}, {@code i0}, {@code el0} vs {@code arr1},
     *     {@code i1}, {@code el1})
     * @return a {@link CodeBlock} containing the serialization logic statements
     */
    private CodeBlock generateSerialization(TypeMirror type, String inputVar, String targetExpression, int depth) {
        CodeBlock.Builder builder = CodeBlock.builder();
        generateSerialization(type, inputVar, targetExpression, depth, builder);
        return builder.build();
    }

    private void generateSerialization(
            TypeMirror type, String inputVar, String targetExpression, int depth, CodeBlock.Builder builder) {
        TypeMirrorWrapper wrappedType = TypeMirrorWrapper.wrap(type);

        // Custom Serializer
        Optional<CustomSerializerInfo> customSerializerOptional = customSerializers.getCustomInfo(type);
        if (customSerializerOptional.isPresent()) {
            CustomSerializerInfo info = customSerializerOptional.get();
            TypeName publicTypeName = configurationClassNameGenerator.publicPropertyClassName(type);
            if (info.isStatic()) {
                builder.addStatement(
                        "$L = $T.serialize(($T) $L, context)",
                        targetExpression,
                        info.serializerClass(),
                        publicTypeName,
                        inputVar);
            } else {
                String fieldName = Strings.uncapitalize(
                                info.serializerClass().getSimpleName().toString())
                        + ConfigurationClassNameGenerator.PROVIDER_SUFFIX;
                builder.addStatement(
                        "$L = this.$L.get().apply(($T) $L, context)",
                        targetExpression,
                        fieldName,
                        publicTypeName,
                        inputVar);
            }
            return;
        }

        // Newtype
        if (typesUtil.isNewtype(type)) {
            TypeElement typeElement = (TypeElement) ((DeclaredType) type).asElement();
            TypeMirror underlyingType = NewtypeUtil.getUnderlyingType(typeElement);
            String accessor = NewtypeUtil.getAccessorName(typeElement);
            String unwrappedVar = inputVar + "_" + depth + "_unwrapped";
            builder.addStatement("$T $L = $L.$L", underlyingType, unwrappedVar, inputVar, accessor);
            generateSerialization(underlyingType, unwrappedVar, targetExpression, depth + 1, builder);
            return;
        }

        // Config type
        if (typesUtil.isConfigType(type)) {
            String saverFieldName = configurationClassNameGenerator.getSerializerProviderFieldName(type);
            builder.addStatement(
                    "$L = this.$L.get().apply(($T) $L, context)",
                    targetExpression,
                    saverFieldName,
                    configurationClassNameGenerator.publicPropertyClassName(type),
                    inputVar);
            return;
        }

        // Generic collections (List, Map)
        if (wrappedType.hasTypeArguments()) {
            String canonicalName = wrappedType.erasure().getQualifiedName();
            if (canonicalName.equals(List.class.getName()) || canonicalName.equals(Set.class.getName())) {
                TypeMirror elementType = wrappedType.getTypeArguments().getFirst();
                String serializeHelper = canonicalName.equals(List.class.getName()) ? "serializeList" : "serializeSet";
                String elVar = "el" + depth;
                String ctxVar = "ctx" + depth;
                String elementTarget = "res" + depth;

                builder.beginControlFlow(
                        "$L = $T.$L($L, context, ($L, $L) ->",
                        targetExpression,
                        CollectionsUtils.class,
                        serializeHelper,
                        inputVar,
                        elVar,
                        ctxVar);
                builder.addStatement("$T $L", DataTree.class, elementTarget);
                generateSerialization(elementType, elVar, elementTarget, depth + 1, builder);
                builder.addStatement("return $L", elementTarget);
                builder.endControlFlow();
                builder.addStatement(")");
                return;
            } else if (canonicalName.equals(Map.class.getName())) {
                var typeArguments = wrappedType.getTypeArguments();
                TypeMirror valueType = typeArguments.get(1);
                String valVar = "val" + depth;
                String ctxVar = "ctx" + depth;
                String valTarget = "mapVal" + depth;

                builder.beginControlFlow(
                        "$L = $T.serializeMap($L, context, ($L, $L) ->",
                        targetExpression,
                        CollectionsUtils.class,
                        inputVar,
                        valVar,
                        ctxVar);
                builder.addStatement("$T $L", DataTree.class, valTarget);
                generateSerialization(valueType, valVar, valTarget, depth + 1, builder);
                builder.addStatement("return $L", valTarget);
                builder.endControlFlow();
                builder.addStatement(")");
                return;
            }
        }

        // Enum type
        if (wrappedType.isEnum()) {
            builder.addStatement(
                    "$L = $T.loadFrom($L == null ? null : $L.name())",
                    targetExpression,
                    DataTreeTransforms.class,
                    inputVar,
                    inputVar);
            return;
        }

        // Basic known serializable types or fallback
        builder.addStatement("$L = $T.loadFrom($L)", targetExpression, DataTreeTransforms.class, inputVar);
    }
}

File: annotation-processor/src/main/scala/me/bristermitten/mittenlib/annotations/compile/SerializationCodeGenerator.scala
package me.bristermitten.mittenlib.annotations.compile

import com.google.inject.Inject
import com.palantir.javapoet.*
import io.toolisticon.aptk.tools.TypeMirrorWrapper
import java.util.{ArrayList, Optional}
import javax.lang.model.element.{Modifier, TypeElement}
import javax.lang.model.`type`.{DeclaredType, TypeMirror}
import me.bristermitten.mittenlib.annotations.ast.{
  AbstractConfigStructure,
  CustomSerializerInfo,
  Property
}
import me.bristermitten.mittenlib.annotations.parser.CustomSerializers
import me.bristermitten.mittenlib.annotations.util.{NewtypeUtil, TypesUtil}
import me.bristermitten.mittenlib.config.{
  CollectionsUtils,
  SerializationContext
}
import me.bristermitten.mittenlib.config.extension.UseObjectMapperSerialization
import me.bristermitten.mittenlib.config.tree.{DataTree, DataTreeTransforms}
import me.bristermitten.mittenlib.util.Strings
import scala.jdk.CollectionConverters.*

import me.bristermitten.mittenlib.annotations.domain.{
  ConfigStructure => DomainConfigStructure
}

import _root_.me.bristermitten.mittenlib.codegen.dsl.*
import _root_.me.bristermitten.mittenlib.codegen.dsl.BlockBuilder.*
import _root_.me.bristermitten.mittenlib.codegen.dsl.given

class SerializationCodeGenerator @Inject() (
    private val configurationClassNameGenerator: ConfigurationClassNameGenerator,
    private val typesUtil: TypesUtil,
    private val customSerializers: CustomSerializers,
    private val methodNames: MethodNames
):

  def isSerializationSupported(ast: AbstractConfigStructure): Boolean =
    ast
      .properties()
      .asScala
      .forall(property => !propertyIsUnserializable(property))

  def isSerializationSupported(ast: DomainConfigStructure): Boolean =
    ast.properties.forall(property => !domainPropertyIsUnserializable(property))

  def getUnsupportedSerializationProperties(
      ast: AbstractConfigStructure
  ): java.util.List[String] =
    val unsupported = new ArrayList[String]()
    for (property <- ast.properties().asScala) {
      if (propertyIsUnserializable(property)) {
        unsupported.add(property.name() + " (" + property.propertyType() + ")")
      }
    }
    unsupported

  def getUnsupportedSerializationProperties(
      ast: DomainConfigStructure
  ): java.util.List[String] =
    val unsupported = new ArrayList[String]()
    for (property <- ast.properties) {
      if (domainPropertyIsUnserializable(property)) {
        unsupported.add(property.name + " (" + property.typeMirror + ")")
      }
    }
    unsupported

  private def domainPropertyIsUnserializable(
      property: me.bristermitten.mittenlib.annotations.domain.Property
  ): Boolean =
    val propertyTypeMirror = property.typeMirror
    if (
      typesUtil.getAnnotation(
        property.element,
        classOf[UseObjectMapperSerialization]
      ) != null
    ) {
      return false
    }
    val wrappedType = TypeMirrorWrapper.wrap(propertyTypeMirror)
    if (typesUtil.isNewtype(propertyTypeMirror)) {
      return false // domain properties don't carry underlying type info easily; treat as serializable
    }
    if (wrappedType.hasTypeArguments) {
      val canonicalName = wrappedType.erasure().getQualifiedName()
      if (
        typesUtil.isCollection(propertyTypeMirror) || canonicalName == classOf[
          java.util.Map[?, ?]
        ].getName
      ) {
        return false // collections are serializable at domain level
      }
      return true
    }
    if (typesUtil.isConfigType(propertyTypeMirror)) {
      return false
    }
    if (isKnownSerializableType(wrappedType)) {
      return false
    }
    customSerializers.getCustomInfo(propertyTypeMirror).isEmpty

  private def propertyIsUnserializable(property: Property): Boolean =
    val propertyTypeMirror = property.propertyType()
    if (
      typesUtil.getAnnotation(
        property.source().element(),
        classOf[UseObjectMapperSerialization]
      ) != null
    ) {
      return false
    }

    val wrappedType = TypeMirrorWrapper.wrap(propertyTypeMirror)

    // Newtypes are serializable if their underlying type is serializable
    if (typesUtil.isNewtype(propertyTypeMirror)) {
      val underlyingType =
        typesUtil.getNewtypeUnderlyingType(propertyTypeMirror)
      return propertyIsUnserializable(
        new Property(
          property.name(),
          underlyingType,
          property.source(),
          property.settings()
        )
      )
    }

    // Check generic types
    if (wrappedType.hasTypeArguments) {
      val canonicalName = wrappedType.erasure().getQualifiedName()
      if (
        typesUtil.isCollection(propertyTypeMirror) || canonicalName == classOf[
          java.util.Map[?, ?]
        ].getName
      ) {
        val typeArguments = wrappedType.getTypeArguments.asScala
        for (typeArgument <- typeArguments) {
          if (
            propertyIsUnserializable(
              new Property(
                property.name(),
                typeArgument,
                property.source(),
                property.settings()
              )
            )
          ) {
            return true
          }
        }
        return false
      }
      return true
    }

    if (typesUtil.isConfigType(propertyTypeMirror)) {
      return false
    }

    if (isKnownSerializableType(wrappedType)) {
      return false
    }

    customSerializers.getCustomInfo(propertyTypeMirror).isEmpty

  def addSerializeMethodsToSaver(
      typeSpecBuilder: TypeSpec.Builder,
      ast: AbstractConfigStructure
  ): Unit =
    for (property <- ast.properties().asScala) {
      val serializeMethod = createSerializeMethodFor(property)
      typeSpecBuilder.addMethod(serializeMethod)
    }

  private def getSerializeParameterType(property: Property): TypeName =
    val typeName =
      configurationClassNameGenerator.publicPropertyClassName(property)
    typeName match
      case parameterizedTypeName: ParameterizedTypeName =>
        val rawType = parameterizedTypeName.rawType()
        if (
          rawType == ClassName.get(classOf[java.util.List[?]]) ||
          rawType == ClassName.get(classOf[java.util.Set[?]]) ||
          rawType == ClassName.get(classOf[java.util.Map[?, ?]])
        ) {
          val typeArguments = parameterizedTypeName
            .typeArguments()
            .asScala
            .map { arg =>
              WildcardTypeName.subtypeOf(arg).asInstanceOf[TypeName]
            }
            .asJava
          ParameterizedTypeName.get(rawType, typeArguments.asScala.toArray*)
        } else {
          typeName
        }
      case _ => typeName

  private def createSerializeMethodFor(property: Property): MethodSpec =
    val methodName = methodNames.getSerializeMethodName(property)
    val propertyType = getSerializeParameterType(property)
    val useObjectMapper = typesUtil.getAnnotation(
      property.source().element(),
      classOf[UseObjectMapperSerialization]
    ) != null

    val builder = MethodSpec
      .methodBuilder(methodName)
      .addJavadoc(
        """Serializes the {@code $L} property into a {@link $T}.
          |
          |@param value the value to serialize
          |@param context the serialization context
          |@return the serialized DataTree representation
          |""".stripMargin,
        property.name(),
        classOf[DataTree]
      )
      .addModifiers(Modifier.PRIVATE)
      .returns(classOf[DataTree])
      .addParameter(propertyType, "value")
      .addParameter(classOf[SerializationContext], "context")

    val dslBlock = BlockBuilder.buildOpen {
      val value = Var("value", TypeRef.of(propertyType))
      val context = Var("context", Types.SerializationContext)

      if (property.settings().isNullable()) {
        ifThen(value.isNull) {
          return_(Expr.staticCall(Types.DataTree, "null_"))
        }
      }

      if (useObjectMapper) {
        return_(
          Expr.staticCall(
            Types.DataTreeTransforms,
            "loadFrom",
            context.call("getMapper").call("map", value)
          )
        )
      } else {
        val result = declare(
          Types.DataTree,
          "result",
          generateSerialization(property.propertyType(), value, 0)
        )
        return_(result)
      }
    }

    builder.addCode(CodeBlockRenderer.render(dslBlock))
    builder.build()

  private def isKnownSerializableType(wrappedType: TypeMirrorWrapper): Boolean =
    if (wrappedType.getTypeElement.isEmpty && wrappedType.isPrimitive) {
      return true
    }

    if (
      typesUtil.getDataTreeType(TypeName.get(wrappedType.unwrap())).isDefined
    ) {
      return true
    }

    if (wrappedType.getQualifiedName == classOf[java.lang.Character].getName) {
      return true
    }
    wrappedType.isEnum

  private def generateSerialization(
      tpe: TypeMirror,
      input: Expr[?],
      depth: Int
  )(using
      BlockBuilder
  ): Expr[?] =
    val wrappedType = TypeMirrorWrapper.wrap(tpe)

    // Custom Serializer
    val customSerializerOptional = customSerializers.getCustomInfo(tpe)
    if (customSerializerOptional.isPresent) {
      val info = customSerializerOptional.get()
      val publicTypeName =
        configurationClassNameGenerator.publicPropertyClassName(tpe)
      val context = Var("context", Types.SerializationContext)
      if (info.isStatic) {
        Expr.staticCall(
          TypeRef.of(ClassName.get(info.serializerClass)),
          "serialize",
          input.cast(TypeRef.of(publicTypeName)),
          context
        )
      } else {
        val fieldName = Strings.uncapitalize(
          info.serializerClass.getSimpleName.toString
        ) + ConfigurationClassNameGenerator.PROVIDER_SUFFIX
        Expr.This
          .field(fieldName)
          .call("get")
          .call("apply", input.cast(TypeRef.of(publicTypeName)), context)
      }
    }
    // Newtype
    else if (typesUtil.isNewtype(tpe)) {
      val typeElement =
        tpe.asInstanceOf[DeclaredType].asElement().asInstanceOf[TypeElement]
      val underlyingType = typesUtil.getNewtypeUnderlyingType(tpe)
      val accessor =
        NewtypeUtil.getAccessorName(typeElement).replaceAll("\\(\\)", "")
      val unwrappedVarName = input match {
        case Var(name, _) => s"${name}_${depth}_unwrapped"
        case _            => s"value_${depth}_unwrapped"
      }
      val unwrappedVar = declare(
        TypeRef.of(TypeName.get(underlyingType)),
        unwrappedVarName,
        input.call(accessor)
      )
      generateSerialization(underlyingType, unwrappedVar, depth + 1)
    }
    // Config type
    else if (typesUtil.isConfigType(tpe)) {
      val saverFieldName =
        configurationClassNameGenerator.getSerializerProviderFieldName(tpe)
      val publicTypeName =
        configurationClassNameGenerator.publicPropertyClassName(tpe)
      val context = Var("context", Types.SerializationContext)
      Expr.This
        .field(saverFieldName)
        .call("get")
        .call("apply", input.cast(TypeRef.of(publicTypeName)), context)
    }
    // Generic collections (List, Map, Set)
    else if (wrappedType.hasTypeArguments) {
      val canonicalName = wrappedType.erasure().getQualifiedName()
      val context = Var("context", Types.SerializationContext)
      if (
        canonicalName == classOf[
          java.util.List[?]
        ].getName || canonicalName == classOf[java.util.Set[?]].getName
      ) {
        val elementType = wrappedType.getTypeArguments.get(0)
        val serializeHelper =
          if (canonicalName == classOf[java.util.List[?]].getName)
            "serializeList"
          else "serializeSet"
        val elVar = Var(s"el$depth", TypeRef.of(TypeName.get(elementType)))
        val ctxVar = Var(s"ctx$depth", Types.SerializationContext)

        val lambdaBlock = BlockBuilder.build {
          val elementTarget = declare(
            Types.DataTree,
            s"res$depth",
            generateSerialization(elementType, elVar, depth + 1)
          )
          return_(elementTarget)
        }

        Expr.staticCall(
          Types.CollectionsUtils,
          serializeHelper,
          input,
          context,
          Expr.Lambda(scala.List(elVar, ctxVar), lambdaBlock)
        )
      } else if (canonicalName == classOf[java.util.Map[?, ?]].getName) {
        val valueType = wrappedType.getTypeArguments.get(1)
        val valVar = Var(s"val$depth", TypeRef.of(TypeName.get(valueType)))
        val ctxVar = Var(s"ctx$depth", Types.SerializationContext)

        val lambdaBlock = BlockBuilder.build {
          val valTarget = declare(
            Types.DataTree,
            s"mapVal$depth",
            generateSerialization(valueType, valVar, depth + 1)
          )
          return_(valTarget)
        }

        Expr.staticCall(
          Types.CollectionsUtils,
          "serializeMap",
          input,
          context,
          Expr.Lambda(scala.List(valVar, ctxVar), lambdaBlock)
        )
      } else {
        throw new IllegalStateException(
          "Unexpected generic type: " + canonicalName
        )
      }
    }
    // Enum type
    else if (wrappedType.isEnum) {
      Expr.staticCall(
        Types.DataTreeTransforms,
        "loadFrom",
        Expr.Ternary(input === Expr.Null, Expr.Null, input.call("name"))
      )
    }
    // Basic fallback
    else {
      Expr.staticCall(Types.DataTreeTransforms, "loadFrom", input)
    }
